src/infrastructure/persistence/PsqlTestRepository.ts

```diff
@@ -67,15 +54,93 @@ export class PsqlTestRepository implements TestRepository {
       return [];
     }
 
-    return testRows.map(
-      (testRow: any) =>
-        new Test(
-          new TestId(testRow.id),
-          new UserId(testRow.userId),
-          new TestTypeId(testRow.testTypeId),
-          undefined,
-          testRow.creationTime
+    return testRows.map(this.createResultsFromRow);
+  }
+
+  private async saveWithoutResults(test: Test) {
+    return await this.db
+      .raw(
+        `
+      insert into "${TEST_TABLE_NAME}" (
+        id,
+        user_id,
+        test_type_id,
+        creation_time
+      )
+      values (
+        :id,
+        :user_id,
+        :test_type_id,
+        :creation_time
+      )
+      on conflict(id) do nothing
```

Comment:
Why are we doing so much raw SQL if we have knex for it?

Reply:
I've just been following the pattern set out in the first commits, where we're using raw for insert.  Not sure why that was originally...

Reply:
Maybe knex doesn't have native syntax for upserts... something to check